File: flutter_hr/HR-And-Payroll-Management-System-Backend/src/main/java/Israt/example/Hr/and/payroll/management/system/backend/entity/Bonus.java
package Israt.example.Hr.and.payroll.management.system.backend.entity;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.math.BigDecimal;
import java.time.LocalDateTime;

@Entity
@Table(name = "bonuses")
@Data
@NoArgsConstructor
@AllArgsConstructor
public class Bonus {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "employee_id", nullable = false)
    private Long employeeId;

    @Column(name = "bonus_type", length = 100)
    private String bonusType;
    @Column(name = "bonus_amount", nullable = false)
    private BigDecimal bonusAmount;

    @Column(name = "percentage_of_salary")
    private Double percentageOfSalary;

    @Column(name = "for_month")
    private Integer forMonth;

    @Column(name = "for_year")
    private Integer forYear;

    @Column(length = 2000)
    private String description;

    @Column(length = 1000)
    private String criteria;

    @Column(length = 50)
    private String status;

    @Column(name = "approved_by")
    private Long approvedBy;

    @Column(name = "approval_date")
    private LocalDateTime approvalDate;

    @Column(length = 1000)
    private String rejectionReason;

    @Column(name = "is_paid")
    private Boolean isPaid = false;

    @Column(name = "paid_date")
    private LocalDateTime paidDate;

    @Column(name = "payment_method", length = 100)
    private String paymentMethod;

    @Column(name = "payment_reference", length = 255)
    private String paymentReference;

    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt = LocalDateTime.now();

    @Column(name = "updated_at")
    private LocalDateTime updatedAt = LocalDateTime.now();

    @Column(name = "created_by")
    private Long createdBy;

    @Column(name = "updated_by")
    private Long updatedBy;

    @PreUpdate
    protected void onUpdate() {
        this.updatedAt = LocalDateTime.now();
    }
}

File: flutter_hr_2/flutter_hr/lib/models/bonus.dart
class Bonus {
  final int? id;
  final int employeeId;
  final String bonusType;
  final double bonusAmount;
  final double? percentageOfSalary;
  final int? forMonth;
  final int? forYear;
  final String? description;
  final String? criteria;
  final String? status;
  final int? approvedBy;
  final String? approvalDate;
  final String? rejectionReason;
  final bool isPaid;
  final String? paidDate;
  final String? paymentMethod;
  final String? paymentReference;
  final String? createdAt;

  Bonus({
    this.id,
    required this.employeeId,
    required this.bonusType,
    required this.bonusAmount,
    this.percentageOfSalary,
    this.forMonth,
    this.forYear,
    this.description,
    this.criteria,
    this.status,
    this.approvedBy,
    this.approvalDate,
    this.rejectionReason,
    this.isPaid = false,
    this.paidDate,
    this.paymentMethod,
    this.paymentReference,
    this.createdAt,
  });

  Bonus copyWith({
    int? id,
    int? employeeId,
    String? bonusType,
    double? bonusAmount,
    double? percentageOfSalary,
    int? forMonth,
    int? forYear,
    String? description,
    String? criteria,
    String? status,
    int? approvedBy,
    String? approvalDate,
    String? rejectionReason,
    bool? isPaid,
    String? paidDate,
    String? paymentMethod,
    String? paymentReference,
    String? createdAt,
  }) {
    return Bonus(
      id: id ?? this.id,
      employeeId: employeeId ?? this.employeeId,
      bonusType: bonusType ?? this.bonusType,
      bonusAmount: bonusAmount ?? this.bonusAmount,
      percentageOfSalary: percentageOfSalary ?? this.percentageOfSalary,
      forMonth: forMonth ?? this.forMonth,
      forYear: forYear ?? this.forYear,
      description: description ?? this.description,
      criteria: criteria ?? this.criteria,
      status: status ?? this.status,
      approvedBy: approvedBy ?? this.approvedBy,
      approvalDate: approvalDate ?? this.approvalDate,
      rejectionReason: rejectionReason ?? this.rejectionReason,
      isPaid: isPaid ?? this.isPaid,
      paidDate: paidDate ?? this.paidDate,
      paymentMethod: paymentMethod ?? this.paymentMethod,
      paymentReference: paymentReference ?? this.paymentReference,
      createdAt: createdAt ?? this.createdAt,
    );
  }

  factory Bonus.fromJson(Map<String, dynamic> json) {
    return Bonus(
      id: json['id'] as int?,
      employeeId: json['employeeId'] as int? ?? 0,
      bonusType: json['bonusType']?.toString() ?? '',
      bonusAmount: json['bonusAmount'] is String 
          ? double.parse(json['bonusAmount']) 
          : (json['bonusAmount'] as num?)?.toDouble() ?? 0.0,
      percentageOfSalary: (json['percentageOfSalary'] as num?)?.toDouble(),
      forMonth: json['forMonth'] as int?,
      forYear: json['forYear'] as int?,
      description: json['description']?.toString(),
      criteria: json['criteria']?.toString(),
      status: json['status']?.toString(),
      approvedBy: json['approvedBy'] as int?,
      approvalDate: json['approvalDate']?.toString(),
      rejectionReason: json['rejectionReason']?.toString(),
      isPaid: json['isPaid'] as bool? ?? false,
      paidDate: json['paidDate']?.toString(),
      paymentMethod: json['paymentMethod']?.toString(),
      paymentReference: json['paymentReference']?.toString(),
      createdAt: json['createdAt']?.toString(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      if (id != null) 'id': id,
      'employeeId': employeeId,
      'bonusType': bonusType,
      'bonusAmount': bonusAmount,
      'percentageOfSalary': percentageOfSalary,
      'forMonth': forMonth,
      'forYear': forYear,
      'description': description,
      'criteria': criteria,
      'status': status,
      'approvedBy': approvedBy,
      'approvalDate': approvalDate,
      'rejectionReason': rejectionReason,
      'isPaid': isPaid,
      'paidDate': paidDate,
      'paymentMethod': paymentMethod,
      'paymentReference': paymentReference,
    };
  }
}
